Order creation: Prevent subscription products to be added to an order

iamgabrielma opened the pull request:
Closes WOOMOB-938

## Description
This PR disallows adding product of type `subscription` and `variableSubscription` to an order via creation or edition flows. Since subscriptions are not currently supported in the app, restricting adding these product types from the get-go will set the expectations right away and remove potential confusion and issues like the one raised in WOOMOB-938

## Changes
We leverage the existing `bookings` implementation and add these two product types as well to the mix. This will disallow the specific product type from being selected and added to the order in both creation and edition flows.

## Testing information
- Install WooCommerce Subscriptions plugin
- Create a subscription product either in wp-admin or via the app. Note that subscription products can still be created via the Product tab, this decision is outside of the scope of this PR
- Navigate to the Orders tab, attempt to create or edit an order adding a subscription product, and confirm that is not possible.

## Screenshots

https://github.com/user-attachments/assets/f3d5598f-f502-4cab-a1cd-e99da98e6b2c

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

iamgabrielma committed:
prevent subscription products to be added to an order

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

iamgabrielma committed:
update release notes

jaclync reviewed (commented):
During testing, I was still able to add a variation from a variable subscription product. Is it expected? I thought tapping on the variable subscription product would be disabled since it's grayed out.

https://github.com/user-attachments/assets/8b7c04f0-39f6-4537-8c35-8158a5e62541

iamgabrielma committed:
Handle variableSubscription case from view

The call to .onTapGesture for variations in the product selector was bypassing the product type check, and navigating the merchant to the variations view, where they could add subscription products to the order

iamgabrielma committed:
simplify, and unit tests

iamgabrielma commented:
> During testing, I was still able to add a variation from a variable subscription product. Is it expected? I thought tapping on the variable subscription product would be disabled since it's grayed out.

Thanks for testing, it seems my own store trolled me and variable subscriptions weren't actually variable but simple 🤦 

<img width="800" height="461" alt="Screenshot 2025-07-31 at 11 21 58" src="https://github.com/user-attachments/assets/b8440e34-3129-4026-a95c-51b0ccd1f2f5" />

The problem relied on variations being handled via a call inside `.tapGesture` in `ProductSelectorView`, which was navigating the merchant to the variations view directly, bypassing the block we set for the parent product when there are no variations. 

I've added a new isSubscription check on https://github.com/woocommerce/woocommerce-ios/pull/15960/commits/debe2a95c9bb9b4f9bf17f3be83f79ecc023451e which now resolves the problem, as well as additional unit tests:

https://github.com/user-attachments/assets/58ecf8b3-b452-4e32-b541-2a42612779e2

jaclync commented on WooCommerce/Classes/ViewRelated/Products/ProductSelector/ProductSelectorView.swift:
```diff
@@ -212,15 +212,20 @@ struct ProductSelectorView: View {
     @ViewBuilder private func createProductRow(rowViewModel: ProductRowViewModel) -> some View {
         if viewModel.isVariableProduct(productOrVariationID: rowViewModel.productOrVariationID),
            configuration.treatsAllProductsAsSimple == false {
+            let isSubscription = viewModel.isSubscriptionProduct(productOrVariationID: rowViewModel.productOrVariationID)
```
❓ Is it possible to use `ProductRowViewModel.SelectedState` == `unsupported` to determine whether a row is selectable/tappable?

jaclync replied:
nit: the `DisclosureIndicator()` can be hidden too if the row isn't tappable

<img width="500" alt="Simulator Screenshot - iPad Air 13-inch (M3) - 2025-07-31 at 16 04 45" src="https://github.com/user-attachments/assets/458b6eb4-b957-4b08-a97a-c6ac36faa84e" />

iamgabrielma replied:
>  Is it possible to use ProductRowViewModel.SelectedState == unsupported to determine whether a row is selectable/tappable?

That's a good point, I didn't see it right away but the existing `selectionEnabled` returns false when `selectedState == .unsupported`, which covers all unsupported product types and we won't need to perform check per-product-type, just expand the list, which is more maintainable in the long term. Updated here:   https://github.com/woocommerce/woocommerce-ios/pull/15960/commits/b419d946d6998bd2e320e3fb585b54e16d7d96e1 . This also comes with the subscription check being used only on tests, so I removed these here https://github.com/woocommerce/woocommerce-ios/pull/15960/commits/866c91c64ef6a3683ff1c374ee910b04e7d02beb

> nit: the DisclosureIndicator() can be hidden too if the row isn't tappable

Good UI suggestion, updated: https://github.com/woocommerce/woocommerce-ios/pull/15960/commits/b5e48c787caf4ac8ff85c59d9bd3b39268173b47 so it's hidden now for non-selectable variable products.

<img width="2266" height="1488" alt="Simulator Screenshot - iPad mini (A17 Pro) - US store - 2025-08-01 at 09 55 21" src="https://github.com/user-attachments/assets/f3b86768-653a-49c9-ba64-9618b079bb53" />

jaclync reviewed (commented):
Some thoughts about the latest update, let me know what you think!

iamgabrielma committed:
simplify check by using existing selectionEnabled

iamgabrielma committed:
hide disclosure indicator for parent variable product when is not selectable

iamgabrielma committed:
remove unused isSubscriptionProduct and associated tests

iamgabrielma committed:
Merge branch 'trunk' into task/WOOMOB-938-prevent-subs-products-from-order-flow

jaclync reviewed (approved):
LGTM! :shipit:
